### PR TITLE
[OFFAPPS-992] Allow groups to send messages

### PR DESCRIPTION
**Pull request by ocke**
Feature from: https://github.com/zendesk/notification_app/issues/40

With this PR it is possible for an admin to allow other people in a group, to also send messages.

1) Added new admin only interface, that allows selecting groups, which are also allowed to send messages.

The cog in the upper right corner brings you to the settings screen.
screen shot 2018-06-29 at 14 57 03

The new settings screen that allows admins to change the 1 setting.
screen shot 2018-06-29 at 14 57 11

2) Users in a group (selected by an admin) can now also send messages!

To make this work seamlessly and instantly, the new settings are also propagated to all agents, so it works without refreshing the browser (and apps).

### CODE REVIEW

File: src/templates/background.html [hunk 1]
```diff
@@ -35,6 +35,20 @@
           })
         })
       })
+
```
Comment (yu-tian113) [inline src/templates/background.html]: top_bar for event?

File: src/javascript/app.js [hunk 2]
```diff
@@ -91,6 +92,26 @@ const app = {
     })
   },
 
+  openSetting: function () {
+    const groups = storage('groups')
+    groups.forEach((group) => {
+      group.isSelected = setting('adminGroups').indexOf(group.id) !== -1
+    })
+
+    app.render('settings', {
```
Comment (yu-tian113) [inline src/javascript/app.js]: 1 line

File: src/javascript/app.js [hunk 3]
```diff
@@ -154,11 +175,12 @@ const app = {
 
   onIncomingMessage: function (message, sender) {
     appReady.then(() => {
-      if (sender.role !== 'admin') return false
+      if (sender.role !== 'admin' && !intersection(storage('myGroupIds'), setting('adminGroups')).length) return false
```
Comment (yu-tian113) [inline src/javascript/app.js]: abstract function: isAuthorized

File: src/templates/inbox.hdbs [hunk 4]
```diff
@@ -1,4 +1,10 @@
 {{#if isAdmin}}
+  <svg class="settings" viewBox="0 0 16 16" width="16" height="16">
+    <path fill="none" stroke="currentColor" d="M13.17 3.07l-1.64.71c-.48-.4-1.03-.72-1.63-.94l-.22-1.79a.511.511 0 0 0-.4-.44C8.86.54 8.44.5 8 .5c-.44 0-.86.04-1.28.11-.22.04-.38.22-.4.44l-.21 1.79c-.6.22-1.15.54-1.63.94l-1.65-.71a.491.491 0 0 0-.58.13c-.55.65-.99 1.4-1.29 2.21-.08.21 0 .44.18.58l1.45 1.08a4.91 4.91 0 0 0 0 1.87l-1.45 1.08c-.18.13-.25.37-.18.58.3.81.74 1.55 1.28 2.2.14.17.38.22.58.13l1.65-.71c.48.4 1.03.72 1.63.94l.21 1.79c.03.22.19.4.4.44.43.07.85.11 1.29.11.44 0 .86-.04 1.28-.11.22-.04.38-.22.4-.44l.21-1.79c.6-.22 1.15-.54 1.63-.94l1.65.71c.2.09.44.04.58-.13.54-.65.98-1.39 1.28-2.2.08-.21 0-.44-.18-.57L13.4 8.95c.07-.32.1-.63.1-.95s-.03-.63-.09-.94l1.45-1.08c.18-.13.25-.37.18-.58-.3-.81-.74-1.55-1.28-2.2a.51.51 0 0 0-.59-.13z"></path>
```
Comment (yu-tian113) [inline src/templates/inbox.hdbs]: can webpack do this?

File: src/templates/settings.hdbs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<label class="c-txt__label c-txt__label--sm">{{t 'label.admin_groups'}}</label>
+{{> datalist name='groups' options=groups }}
+
+<p class="helptext">
+  {{t 'label.admin_groups_helptext'}}
+</p>
+
+<div style="height: 300px"></div>
```
Comment (yu-tian113) [inline src/templates/settings.hdbs]: errrrrrr....

File: src/templates/inbox.hdbs (hunk 4 above)
Comment (ocke) [reply src/templates/inbox.hdbs]: I tried, but i can't get it to work.

File: src/javascript/app.js [hunk 6]
```diff
@@ -92,16 +102,52 @@ const app = {
...
+    getDatalist('groups').on('open_menu', (event, datalist) => {
+      const $datalist = datalist.$datalist()
+      const $menu = $datalist.find('.menu_options')
+      // datalist top + menu top + menu height + little bottom margin
+      const newHeight = $datalist[0].offsetTop + $menu[0].offsetTop + $menu[0].offsetHeight + 6
+      appResize(newHeight)
+    })
+
+    getDatalist('groups').on('close_menu', () => {
+      // Delay because appResize triggers a draw event and thus never runs the other events?!
```
Comment (yu-tian113) [inline src/javascript/app.js]: I think this comment is not accurate as the root cause you found? 
maybe `delay the resize till click event is fired`

**Comment by m-lai** [comment]
Hey @ocke , can we please:

1. Switch the tags in the multiselect to the smaller sized ones to reduce the amount of space they take up for better legibility in a small screen? (also note the padding is less than what you have) https://cl.ly/1T3e0V3o0P2K

2. Make sure the cog icon color is gray-600 by default, gray-700 on hover?

cc @sostopher The gear icon should indeed be the outline version as Garden has moved over to using new icons now and stroke is the default version for all new features: https://cl.ly/1q1d1Q2N430Z

Thanks

**Comment by sostopher** [comment]
@ocke Can you please add the ticket ref in the PR title? (OFFAPPS-978)? Thanks.

**Pull request merged**